solidity/contracts/converter/types/standard-pool/StandardPoolConverter.sol

```diff
@@ -0,0 +1,1386 @@
+    }
+
+    /**
+     * @dev updates the recent average rate if needed
+     */
+    function updateRecentAverageRate() internal {
+        uint256 averageRateInfo1 = averageRateInfo;
+        uint256 averageRateInfo2 = calcRecentAverageRate(averageRateInfo1);
+        if (averageRateInfo1 != averageRateInfo2) {
+            averageRateInfo = averageRateInfo2;
+        }
+    }
+
+    /**
+     * @dev returns the recent average rate of 1 reserve token 0 in reserve token 1 units
+     *
+     * @param _averageRateInfo a local copy of the `averageRateInfo` state-variable
+     * @return recent average rate between the reserves
+     */
+    function calcRecentAverageRate(uint256 _averageRateInfo) internal view returns (uint256) {
+        // get the previous average rate and its update-time
+        uint256 prevAverageRateT = decodeAverageRateT(_averageRateInfo);
+        uint256 prevAverageRateN = decodeAverageRateN(_averageRateInfo);
+        uint256 prevAverageRateD = decodeAverageRateD(_averageRateInfo);
+
+        // get the elapsed time since the previous average rate was calculated
+        uint256 currentTime = time();
+        uint256 timeElapsed = currentTime - prevAverageRateT;
+
+        // if the previous average rate was calculated in the current block, the average rate remains unchanged
+        if (timeElapsed == 0) {
+            return _averageRateInfo;
+        }
+
+        // get the current rate between the reserves
+        (uint256 currentRateD, uint256 currentRateN) = reserveBalances();
+
+        // if the previous average rate was calculated a while ago or never, the average rate is equal to the current rate
+        if (timeElapsed >= AVERAGE_RATE_PERIOD || prevAverageRateT == 0) {
+            (currentRateN, currentRateD) = Math.reducedRatio(currentRateN, currentRateD, MAX_UINT112);
+            return encodeAverageRateInfo(currentTime, currentRateN, currentRateD);
+        }
+
+        uint256 x = prevAverageRateD.mul(currentRateN);
+        uint256 y = prevAverageRateN.mul(currentRateD);
+
+        // since we know that timeElapsed < AVERAGE_RATE_PERIOD, we can avoid using SafeMath:
+        uint256 newRateN = y.mul(AVERAGE_RATE_PERIOD - timeElapsed).add(x.mul(timeElapsed));
+        uint256 newRateD = prevAverageRateD.mul(currentRateD).mul(AVERAGE_RATE_PERIOD);
+
+        (newRateN, newRateD) = Math.reducedRatio(newRateN, newRateD, MAX_UINT112);
+        return encodeAverageRateInfo(currentTime, newRateN, newRateD);
+    }
+
+    /**
+     * @dev increases the pool's liquidity and mints new shares in the pool to the caller
+     * note that prior to version 28, you should use 'fund' instead
+     *
+     * @param _reserveTokens   address of each reserve token
+     * @param _reserveAmounts  amount of each reserve token
+     * @param _minReturn       token minimum return-amount
+     *
+     * @return amount of pool tokens issued
+     */
+    function addLiquidity(
+        IERC20Token[2] memory _reserveTokens,
```

Comment:
I think that declaring them as `calldata` will result in slightly cheaper gas costs (since `calldata` args are immutable)

Reply:
I don't understand.
First, they are not immutable at this point, because function `VerifyInput` inverts the order if necessary.
So at most, we can apply it in each one of the underlying (private) functions.
Second - this is a broad change, which should be applied globally, so why only on this function?